src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -14,44 +18,51 @@ def list(self):
         osds = self.get_osd_map()
 
         # Extending by osd stats information
-        for s in mgr.get('osd_stats')['osd_stats']:
-            osds[str(s['osd'])].update({'osd_stats': s})
+        for stat in mgr.get('osd_stats')['osd_stats']:
+            if stat['osd'] in osds:
+                osds[stat['osd']]['osd_stats'] = stat
 
         # Extending by osd node information
         nodes = mgr.get('osd_map_tree')['nodes']
-        osd_tree = [(str(o['id']), o) for o in nodes if o['id'] >= 0]
-        for o in osd_tree:
-            osds[o[0]].update({'tree': o[1]})
+        for node in nodes:
+            if node['type'] == 'osd' and node['id'] in osds:
+                osds[node['id']]['tree'] = node
 
         # Extending by osd parent node information
-        hosts = [(h['name'], h) for h in nodes if h['id'] < 0]
-        for h in hosts:
-            for o_id in h[1]['children']:
-                if o_id >= 0:
-                    osds[str(o_id)]['host'] = h[1]
+        for host in [n for n in nodes if n['type'] == 'host']:
+            for osd_id in host['children']:
+                if osd_id >= 0 and osd_id in osds:
+                    osds[osd_id]['host'] = host
 
         # Extending by osd histogram data
-        for o_id in osds:
-            o = osds[o_id]
-            o['stats'] = {}
-            o['stats_history'] = {}
-            osd_spec = str(o['osd'])
-            for s in ['osd.op_w', 'osd.op_in_bytes', 'osd.op_r', 'osd.op_out_bytes']:
-                prop = s.split('.')[1]
-                o['stats'][prop] = CephService.get_rate('osd', osd_spec, s)
-                o['stats_history'][prop] = CephService.get_rates('osd', osd_spec, s)
+        for osd_id, osd in osds.items():
+            osd['stats'] = {}
+            osd['stats_history'] = {}
+            osd_spec = str(osd_id)
+            if 'osd' not in osd:
+                continue
+            for stat in ['osd.op_w', 'osd.op_in_bytes', 'osd.op_r', 'osd.op_out_bytes']:
+                prop = stat.split('.')[1]
+                osd['stats'][prop] = CephService.get_rate('osd', osd_spec, stat)
+                osd['stats_history'][prop] = CephService.get_rates('osd', osd_spec, stat)
             # Gauge stats
-            for s in ['osd.numpg', 'osd.stat_bytes', 'osd.stat_bytes_used']:
-                o['stats'][s.split('.')[1]] = mgr.get_latest('osd', osd_spec, s)
+            for stat in ['osd.numpg', 'osd.stat_bytes', 'osd.stat_bytes_used']:
+                osd['stats'][stat.split('.')[1]] = mgr.get_latest('osd', osd_spec, stat)
 
         return list(osds.values())
 
-    def get_osd_map(self):
-        osds = {}
-        for osd in mgr.get('osd_map')['osds']:
+    @staticmethod
+    def get_osd_map(svc_id=None):
+        # type: (Union[int, None]) -> Dict[int, Union[Dict[str, Any], Any]]
+        def add_id(osd):
             osd['id'] = osd['osd']
-            osds[str(osd['id'])] = osd
-        return osds
+            return osd
+        resp = {
```

Comment:
I suggest to indent it as done by [yapf](https://github.com/google/yapf) (a Python beautifier i use in Atom/PyCharm).
```
resp = {
    osd['osd']: add_id(osd)
    for osd in mgr.get('osd_map')['osds'] if svc_id is None or (
        svc_id is not None and osd['osd'] == int(svc_id))
}
```

Reply:
```
$ cat /tmp/blah.py     
class A(object):
    def f(self):
        resp = {
            osd['osd']: add_id(osd)
            for osd in mgr.get('osd_map')['osds'] if svc_id is None or (
            svc_id is not None and osd['osd'] == int(svc_id))
        }
        return resp
$ yapf3 /tmp/blah.py
class A(object):
    def f(self):
        resp = {
            osd['osd']: add_id(osd)
            for osd in mgr.get('osd_map')['osds'] if svc_id is None or (
                svc_id is not None and osd['osd'] == int(svc_id))
        }
        return resp
$ mv ~/.config/yapf/nostyle ~/.config/yapf/style 
$ cat ~/.config/yapf/style
[style]
column_limit = 100
user@home ~/src/ceph-1 (wip-pna-fix-osd*) $ yapf3 /tmp/blah.py                            
class A(object):
    def f(self):
        resp = {
            osd['osd']: add_id(osd)
            for osd in mgr.get('osd_map')['osds']
            if svc_id is None or (svc_id is not None and osd['osd'] == int(svc_id))
        }
        return resp
$ 
```

Reply:
```python
if svc_id is None or (svc_id is not None and osd['osd'] == int(svc_id))
```
you don't need `svc_id is not None`, because python won't continue evaluating the whole expression, if it is already `True`.

```
In [2]: True or 1/0
Out[2]: True
```